src/main/mondrian/rolap/SqlContextConstraint.java

```diff
@@ -107,11 +119,28 @@ public static boolean isValidContext(
 
         // we can not handle all calc members in slicer. Calc measure and some
         // like aggregates are exceptions
+        int maxConstraints =
+            MondrianProperties.instance().MaxConstraints.get();
         Member[] members = context.getMembers();
-        for (int i = 1; i < members.length; i++) {
-            if (members[i].isCalculated()
-                && !SqlConstraintUtils.isSupportedCalculatedMember(members[i]))
-            {
+        for (int i = 0; i < members.length; i++) {
+            if (!(members[i].isCalculated())) {
+                // nothing more to check
+                continue;
+            }
+            // dealing with calculated members
+            if (!SqlConstraintUtils.isSupportedCalculatedMember(members[i])) {
+                return false;
+            }
+            // http://jira.pentaho.com/browse/MONDRIAN-1999
+            int calculatedMembersQuantity =
+                SqlConstraintUtils
+                    .expandExpressions(members[i], null, context).size();
+            if (calculatedMembersQuantity > maxConstraints) {
```

Comment:
This is not quite right if the calculated member has a compound aggregate, like Aggregate( CrossJoin( set1, set2))
As it turns out native evaluation mishandles that elsewhere as well, though.   I just logged MONDRIAN-2441 for that.
Whatever we do to fix 2441 will need to update how this count is calculated as well.  I've added a comment to the case.